resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala

```diff
@@ -16,256 +16,130 @@
-      .provideInitContainerBundle(maybeSubmittedResourceIdentifiers.map(_.ids()),
-        resolvedSparkJars ++ resolvedSparkFiles)
-
-    val podWithInitContainer = initContainerBundler.map(
-      _.sparkPodInitContainerBootstrap
-        .bootstrapInitContainerAndVolumes(driverContainer.getName, basePod))
-      .getOrElse(basePod)
-    val sparkConfWithExecutorInit = initContainerBundler.map(
-      _.executorInitContainerConfiguration
-        .configureSparkConfForExecutorInitContainer(sparkConf))
-      .getOrElse(sparkConf)
-    val credentialsMounter = kubernetesCredentialsMounterProvider
-        .getDriverPodKubernetesCredentialsMounter()
-    val credentialsSecret = credentialsMounter.createCredentialsSecret()
-    val podWithInitContainerAndMountedCreds = credentialsMounter.mountDriverKubernetesCredentials(
-      podWithInitContainer, driverContainer.getName, credentialsSecret)
-    val resolvedSparkConf = credentialsMounter.setDriverPodKubernetesCredentialLocations(
-        sparkConfWithExecutorInit)
-    if (resolvedSparkJars.nonEmpty) {
-      resolvedSparkConf.set("spark.jars", resolvedSparkJars.mkString(","))
-    }
-    if (resolvedSparkFiles.nonEmpty) {
-      resolvedSparkConf.set("spark.files", resolvedSparkFiles.mkString(","))
-    }
-    resolvedSparkConf.setIfMissing(KUBERNETES_DRIVER_POD_NAME, kubernetesDriverPodName)
-    resolvedSparkConf.set("spark.app.id", kubernetesAppId)
-    resolvedSparkConf.set(KUBERNETES_EXECUTOR_POD_NAME_PREFIX, kubernetesResourceNamePrefix)
-    // We don't need this anymore since we just set the JVM options on the environment
-    resolvedSparkConf.remove(org.apache.spark.internal.config.DRIVER_JAVA_OPTIONS)
-    val resolvedLocalClasspath = containerLocalizedFilesResolver
-      .resolveSubmittedAndRemoteSparkJars()
-    val resolvedDriverJavaOpts = resolvedSparkConf.getAll.map {
-      case (confKey, confValue) => s"-D$confKey=$confValue"
-    }.mkString(" ") + driverJavaOptions.map(" " + _).getOrElse("")
-    val resolvedDriverPodBuilder = podWithInitContainerAndMountedCreds.editSpec()
-      .editMatchingContainer(new ContainerNameEqualityPredicate(driverContainer.getName))
-        .addNewEnv()
-          .withName(ENV_MOUNTED_CLASSPATH)
-          .withValue(resolvedLocalClasspath.mkString(File.pathSeparator))
-          .endEnv()
-        .addNewEnv()
-          .withName(ENV_DRIVER_JAVA_OPTS)
-          .withValue(resolvedDriverJavaOpts)
-          .endEnv()
-        .endContainer()
-      .endSpec()
-    val driverPodFileMounter = initContainerComponentsProvider.provideDriverPodFileMounter()
-    val resolvedDriverPod = pythonResource.map {
-      p => p.driverPodWithPySparkEnvs(
-        driverPodFileMounter,
-        resolvedPrimaryPySparkResource,
-        resolvedPySparkFiles.mkString(","),
-        driverContainer.getName,
-        resolvedDriverPodBuilder
-      )}.getOrElse(resolvedDriverPodBuilder.build())
+      .build()
     Utils.tryWithResource(
```

Comment:
I'm actually not sure how much this buys us in the end. Extracting the logic to a separate module makes `Client#run()` trivial. It would look something like this:

```
  def run(): Unit = {
    var currentDriverSpec = KubernetesDriverSpec.initialSpec(submissionSparkConf)
    // submissionSteps contain steps necessary to take, to resolve varying
    // client arguments that are passed in, created by orchestrator
    for (nextStep <- submissionSteps) {
      currentDriverSpec = nextStep.configureDriver(currentDriverSpec)
    }
    driverRunner.runDriver(currentDriverSpec)
  }
```

The problem with this kind of abstraction is that it just takes the bulk of complexity from one class and moves it to another. We don't get much extra testability out of this as a result because we're just moving the test logic from one test to another. This feels at least to me like "abstraction for the sake of abstraction". What do you think? cc @ifilonenko also for thoughts.